core/tests/vm-benchmark/benches/instructions.rs

```diff
@@ -0,0 +1,198 @@
+    reporter::{BenchmarkOutput, BenchmarkReporter, Reporter},
+    AccessSummary, Bencher, BenchmarkId,
+};
+
+fn benchmarks_for_vm<VM: BenchmarkingVmFactory>(bencher: &mut Bencher) {
+    bencher.bench(
+        BenchmarkId::new("init", VM::LABEL.as_str()),
+        BenchmarkingVm::<VM>::default,
+    );
+
+    for bytecode in BYTECODES {
+        bencher.bench_with_capture(
+            BenchmarkId::new(bytecode.name, VM::LABEL.as_str()),
+            |capture| {
+                let mut vm = yab::black_box(BenchmarkingVm::<VM>::default());
+                let tx = yab::black_box(bytecode.deploy_tx());
+                capture.measure(|| vm.run_transaction(&tx));
+            },
+        );
+    }
+}
+
+/// Reporter that pushes cachegrind metrics to Prometheus.
+#[derive(Debug)]
+struct MetricsReporter {
+    _runtime: Option<PrometheusRuntime>,
+}
+
+impl Default for MetricsReporter {
+    fn default() -> Self {
+        Self {
+            _runtime: PrometheusRuntime::new(),
+        }
+    }
+}
+
+impl Reporter for MetricsReporter {
+    fn new_benchmark(&mut self, id: &BenchmarkId) -> Box<dyn BenchmarkReporter> {
+        Box::new(MetricsBenchmarkReporter(id.clone()))
+    }
+}
+
+#[derive(Debug)]
+struct MetricsBenchmarkReporter(BenchmarkId);
+
+impl BenchmarkReporter for MetricsBenchmarkReporter {
+    fn ok(self: Box<Self>, output: &BenchmarkOutput) {
+        #[derive(Debug, Metrics)]
+        #[metrics(prefix = "vm_cachegrind")]
+        struct VmCachegrindMetrics {
+            #[metrics(labels = ["benchmark"])]
+            instructions: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l1_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l2_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            ram_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            cycles: LabeledFamily<String, Gauge<u64>>,
+        }
+
+        #[vise::register]
+        static VM_CACHEGRIND_METRICS: vise::Global<VmCachegrindMetrics> = vise::Global::new();
+
+        let id = self.0.to_string();
+        VM_CACHEGRIND_METRICS.instructions[&id].set(output.stats.total_instructions());
+        if let Some(&full) = output.stats.as_full() {
+            let summary = AccessSummary::from(full);
+            VM_CACHEGRIND_METRICS.l1_accesses[&id].set(summary.l1_hits);
+            VM_CACHEGRIND_METRICS.l2_accesses[&id].set(summary.l3_hits);
+            VM_CACHEGRIND_METRICS.ram_accesses[&id].set(summary.ram_accesses);
+            VM_CACHEGRIND_METRICS.cycles[&id].set(summary.estimated_cycles());
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Comparison {
+    current_cycles: u64,
+    prev_cycles: Option<u64>,
+}
+
+impl Comparison {
+    fn percent_difference(a: u64, b: u64) -> f64 {
+        ((b as f64) - (a as f64)) / (a as f64) * 100.0
```

Comment:
Better to subtract first, then convert to floating point.

Reply:
The difference may be negative. Do you want to convert `a` and `b` to `i64` before subtraction? If so, IMO, it's an unnecessary complication given that this place isn't performance-critical.

Reply:
You are correct that it would be a bit ugly to convert to signed first. The reason I suggested is that subtraction is perfectly accurate for integers but not for floats.

f64 is accurate enough for all simple math, though, so it hardly matters. We're not going to bother to do the whole thing in fixed point, even though that would be more accurate in extreme cases.